include/oneapi/tbb/detail/_task_handle.h

```diff
@@ -136,32 +136,52 @@ class task_dynamic_state {
 };
 #endif // __TBB_PREVIEW_TASK_GROUP_EXTENSIONS
 
+template <typename DerivedType>
+void finalize_base(task_handle_task* p, d1::small_object_allocator& alloc, const d1::execution_data* ed) {
+    if (ed) {
+        alloc.delete_object(static_cast<DerivedType*>(p), *ed);
+    } else {
+        alloc.delete_object(static_cast<DerivedType*>(p));
+    }
+}
+
 class task_handle_task : public d1::task {
-    std::uint64_t m_version_and_traits{};
+    // The pointer to the instantiation of finalize_base with the concrete derived type class for correct finalization
+    // Reuse the first std::uint64_t field (was m_version_and_traits) to preserve backward compatibility
+    using finalize_func_type = void (*)(task_handle_task*, d1::small_object_allocator&, const d1::execution_data*);
+    std::uint64_t m_finalize_func;
+    static_assert(sizeof(finalize_func_type) <= sizeof(m_finalize_func), "Cannot fit finalize pointer into std::uint64_t");
+
     d1::wait_tree_vertex_interface* m_wait_tree_vertex;
     d1::task_group_context& m_ctx;
     d1::small_object_allocator m_allocator;
 #if __TBB_PREVIEW_TASK_GROUP_EXTENSIONS
     std::atomic<task_dynamic_state*> m_dynamic_state;
 #endif
 public:
+
     void finalize(const d1::execution_data* ed = nullptr) {
-        if (ed) {
-            m_allocator.delete_object(this, *ed);
+        finalize_func_type finalize_func = reinterpret_cast<finalize_func_type>(m_finalize_func);
+        if (finalize_func != nullptr) {
+            // If the finalize function is set for the current instantiation - use it
+            (*finalize_func)(this, m_allocator, ed);
         } else {
-            m_allocator.delete_object(this);
+            // Otherwise, the object was compiled with the old version of the library
+            // preserve the previous behavior to avoid undefined behavior
+            finalize_base<task_handle_task>(this, m_allocator, ed);
```

Comment:
I wonder if it would be better to just let this memory leak if we cannot deallocate it properly. Deallocating memory with a wrong size scares me more than a leak - unless we know that `small_object_allocator` deals well with that.

Reply:
I agree. Replaced with a call to task destructor without deallocating the memory in else branch.